Docker example

dilyn-corner opened the pull request:
## Pull Request introduction

Please fill out the template!

- [ ] This PR fixes issue #<foo>
- [x] I have signed the [Canonical contributor license agreement](https://ubuntu.com/legal/contributors)
- [x] I have followed the [Style Guide](/README.md#Contributing)


## Pull Request description

Please provide a brief description of what your PR attempts to accomplish and
how it does so. This would include resolutions of previous `TODO`s, the addition
of more tests, implementing something in the workflows, adding something to the
documentation, etc.

Additionally, if this PR includes a new example snap being added to the
repository, explain why such an example snap is a good one to be added.

This PR adds an example snap which shows how to use the Docker snap. It provides various examples like using the `docker{-compose}` commands directly as well as interacting with the REST API. It also shows how one can use volumes between their containers, setting up very simple containers (web servers) or complicated ones (EdgeX), and also shows how to pre-seed images to the device so that it may not require an internet connection to function.


## Pull Request elaboration

Please provide a brief description of any `TODO`s you have introduced in this
PR, as well as any tests you've added for your work (if you have not implemented
tests, these should be marked as `TODO`s).

Currently there are no tests for this example and that is a potential TODO. However, some very basic tests have been added to the GitHub workflow.

- [x] I have documented any new functions or features either in code or in the
  example's README
- [ ] I have added relevant tests or marked `TODO`s for what successful tests
  look like

dilyn-corner committed:
using-docker: new example of using the Docker snap

The using-docker example provides a snap which contains a collections of
scripts and files to show a minimum-viable way of orchestrating
containers on a target Ubuntu Core device using the Docker snap.
Essentially, the example acts as a wrapper around Docker to automate its
use on a device.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
using-docker: add README explaining the components

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
github: add tests for using-docker/

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
using-docker: add a gitignore

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
README: add new example explainer

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

atomcult commented on using-docker/src/usr/bin/tocker_volume.sh:
Since this is being run as a systemd daemon, should we be `--detach`ing the process?

dilyn-corner replied:
If we detach don't we lose console logging entirely?

atomcult replied:
Sorry, I worded that a bit ambiguously.
Currently, in `tocker_volume.sh` docker is being invoked with `docker run -d ...` which, as you said, I think we don't want.

dilyn-corner replied:
Sorry, I actually can't read and refuse to learn.

atomcult commented on using-docker/snap/hooks/configure:
It might be preferable to use `docker load -i "$image"` here and in the `connect-plug-docker` hook.
`docker import` will import a rootfs tarball, but ignore any metadata, so you have to reference the imported image with its hash (which is effectively random). `docker load` takes metadata into account.

dilyn-corner replied:
That's a fair change! 

Too many subcommands.

atomcult commented on using-docker/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,145 @@
+      # If we already have images in the project directory, assume that is all
+      # of the images we want to include
+      for image in images/*.docker; do
+        [ -e ${CRAFT_PROJECT_DIR}/images/*.docker ] || {
```
I think this if-statement should probably be outside of the loop. Also, `test` will freak out if it receives more than one argument (which it will, if there's more than one `.docker` file.

Additionally, there's no logic to install `$image`.

A good way to do this might be to install every docker image from `${CRAFT_PROJECT_DIR}/images` to the appropriate location. Then check if there are any images in the folder, and, if not,
`docker save -o "${CRAFT_INSTALL_PART}/usr/share/images/hello-world.docker" hello-world`

dilyn-corner replied:
> Also, test will freak out if it receives more than one argument (which it will, if there's more than one .docker file.

This is why `[` cannot be outside of the loop and instead we iterate over the images which exist; if there are no objects for the glob to expand to the loop runs once trivially, so we avoid that by using a test; if there aren't actually objects there, *then* fetch some objects. The logic is essentially to avoid the problem of unknown names.

> Additionally, there's no logic to install $image.

Yeah, that's a typo 😅 

> ... Then check if there are any images in the folder ...

This may be better... Though I think it's just reversing the logic order?

atomcult replied:
Hmm.. maybe I'm misunderstanding what's going on here, but won't
`[ -e ${CRAFT_PROJECT_DIR}/images/*.docker ]` match every file ending in `.docker` each iteration of the loop?

We could do something like
```sh
# try installing any existing images
_image_count=0
for _image in images/*.docker; do
  # do all the stuff to install the docker image
  : $(( _image_count += 1 ))
done

# if there are no existing images, install the default
[ "$_image_count" -gt 0 ] || {
  # start up docker, pull hello-world, and install
}
```

dilyn-corner replied:
Oh shit yes; THAT (`[`) should be on `$image` >.<
Your suggestion seems fine though.

dilyn-corner replied:
este

```
    override-build: |
      craftctl default
      # If we already have images in the project directory, assume that is all
      # of the images we want to include
      for _image in ${CRAFT_PROJECT_DIR}/images/*.docker; do
        [ -e $_image ] || {
          systemctl is-active -q docker || systemctl start docker

          docker pull hello-world || sleep 30 && docker pull hello-world
          docker save -o ./hello-world.docker hello-world

          install -Dm644 hello-world.docker \
            ${CRAFT_PART_INSTALL}/usr/share/images/hello-world.docker
        }

        install -Dm644 $_image \
          ${CRAFT_PART_INSTALL}/usr/share/images/${_image##*/}
      done
```
      
oder este
      
```
    override-build: |
      craftctl default
      
      _image_count=0
      for _image in ${CRAFT_PROJECT_DIR}/images/*.docker; do
        install -Dm644 $_image ${CRAFT_PART_INSTALL}/usr/share/images/${_image##*/}
        : $(( _image_count += 1))
      done

      [ $_image_count -gt 0 ] || {
        systemctl is-active -q docker || systemctl start docker

        docker pull hello-world || sleep 30 && docker pull hello-world

        mkdir -p ${CRAFT_PART_INSTALL}/usr/share/images
        docker save -o ${CRAFT_PART_INSTALL}/usr/share/images/hello-world.docker
      }
      
```

atomcult replied:
I think for the first one, the existence of `_image` is implied by its enumeration in the for loop, so the statements after the `||` are unreachable.

dilyn-corner replied:
Try it 👀

atomcult commented on using-docker/src/usr/bin/tocker_rest.sh:
should these URLs be `http://localhost/v1.41/...`?

dilyn-corner replied:
Good catch!

atomcult commented on using-docker/src/usr/bin/tocker_run.sh:
It's rare that you'd run a docker container with just `docker run <image>`. It might be good to convert this to `docker run "$@"` or even just replace this script with `command: docker-bin/bin/docker` in the `snapcraft.yaml` to allow arbitrary commands.

dilyn-corner replied:
I don't believe you can point the `command` to a non-existent file at snap build-time, but I suppose we can allow many args

atomcult reviewed (commented):
Looks pretty good!
I haven't actually run the snap, but I left some comments for the things that stuck out to me.

dilyn-corner committed:
tocker_rest.sh: fix Docker REST API URLs

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
tocker_run.sh: `docker run` can accept many args

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
tocker_volume.sh: don't detach the process

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
using-docker: s/import/load --input/ snap/hooks/*

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

atomcult commented on using-docker/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,145 @@
+      # of the images we want to include
+      for image in images/*.docker; do
+        [ -e ${CRAFT_PROJECT_DIR}/images/*.docker ] || {
+          systemctl is-active -q docker || systemctl start docker && sleep 30
```
Could we do something like this? Or does systemctl report a service as started before it's actually available?
```sh
systemctl start docker

_wait=0
while ! systemctl is-active -q docker; do
  [ "$_wait" -lt 30 ] || break
  sleep 1
  : $(( _wait += 1 ))
done
```

dilyn-corner replied:
> does systemctl report a service as started before it's actually available

what do you think 🙃

atomcult replied:
yeah, figures :smiling_face_with_tear:

dilyn-corner replied:
We could send trivial requests to the socket and do some cute micro-sleeps but there's no good way of knowing when the docker daemon is actually available.

dilyn-corner committed:
using-docker: use compose as a docker subcommand

docker-compose as a python script was deprecated long ago. By using a
layout for a file provided by the docker snap, we will be able to
correctly use the compose plugin (and, theoretically, all other provided
plugins).

Python should no longer be a requirement for proper functionality.

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner committed:
using-docker: add NOTE that content-interface may one day be deprecated

Signed-off-by: Dilyn Corner <dilyn.corner@canonical.com>

dilyn-corner commented:
Thanks for the feedback! Tests are passing on my end, local testing has also succeeded. It seems like it works great!
